BU-3740 Flag when json recovered from truncated

samjewell opened the pull request:
### ~~Review after #2~~

- Sub-task/issue: https://bridgeu.atlassian.net/browse/BU-3740
- Scope document: https://paper.dropbox.com/doc/BU-3585-Rebuild-Mixpanel-Data-Zs1uFDF87kH1Zv3P7kBjt

---

This pull request adds a key value pair `_truncated: true` to any JSON Hashes (or the string `"_truncated"` to any Arrays) that were truncated and `recover!`ed from.

samjewell committed:
Work in progress - flag when JSON parse had to recover truncated

samjewell committed:
Add tests and cleanup WIP. Add `"_truncated": true` to Hashes, and…

`"_truncated"` to Arrays

gareth commented on lib/rjson/handler.rb:
```diff
@@ -30,7 +32,17 @@ def push o
+        elsif output.is_a? Array
+          output.push "_truncated"
+        else
+          throw "JSON doc is a single, truncated scalar"
```
This is an unconventional use of `throw` (or at least I'm not clear on its purpose). What's the intention here?

samjewell replied:
Hmm - I was thinking that the valid json documents `"foo"` and `23` would hit this `else` block, and at that point we would want to throw an error, as those scalar values will probably be truncated, but won't have been flagged as `_truncated`.

However, the entire JSON parser crashes out on `"foo"` or `23` with an error as follows, so I think I'll just remove this for now.
```
[3] pry(main)> RJSON.load '"foo"'
NoMethodError: undefined method `first' for :root:Symbol
from /Users/samjewell/bu/rjson/lib/rjson/handler.rb:35:in `result'
[4] pry(main)> RJSON.load '12'
NoMethodError: undefined method `first' for :root:Symbol
from /Users/samjewell/bu/rjson/lib/rjson/handler.rb:35:in `result'
```

gareth commented on lib/rjson/handler.rb:
```diff
@@ -45,5 +57,29 @@ def process type, rest
+    # Recover an invalid parse tree by dropping items that we're not certain are
+    # fully recoverable
+    def recover!
+      @truncated = true
```
I'm a little wary that `recover!` is called on *any* error, not just the truncated ones. For the Mixpanel work we're sure that this is the only error there's going to be – because the fields *are* based on valid JSON – but I don't know that we can get away with that in this general-purpose fork.

It significantly changes the cases where this library can be used, and if we're not going to be more generic then we at least have to document that.

gareth reviewed (commented):
I'll avoid approve/rejecting this until the parent PR is ready, but there are a couple of comments

samjewell committed:
Remove `else-throw` branch which is never hit.

It was envisaged for `"foo"` or `23` json docs, but those crash out
Much earlier in the parser anyway, and never hit this line of code.

samjewell committed:
Add documentation and a TODO: "…only flag `_truncated` for that error…"

samjewell commented:
@gareth thanks for the comments-  I've tried to address both of them, so this is ready for your second review now 📦

gareth reviewed (approved):
I'm not 100% convinced about the value of adding to a truncated array – the underscore prefix for property names is a lot more common a convention, it doesn't really carry across to arrays. But there's no other way to represent that data in JSON, so ok.
